modules/gui/src/com/haulmont/cuba/gui/data/impl/AbstractCollectionDatasource.java

```diff
@@ -64,8 +65,7 @@
     protected Map<String, Object> savedParameters;
     protected Throwable dataLoadError;
     protected boolean listenersSuspended;
-    protected Operation lastCollectionChangeOperation;
-    protected List<T> lastCollectionChangeItems;
+    protected final LinkedList<Pair<Operation, List<T>>> suspendedEvents = new LinkedList<>();

```

Comment:
What about using a class like SuspendedEvent instead of Pair? Together with LinkedList the usage of the current construct looks horrible - getFirst().getFirst()

Reply:
OK, no problem

Reply:
I've replaced Pair with CollectionChangeEvent.

Actually I've planned to do this in the first place, but for some reason I've thought that the event itself have unmodifiable collection. 😬